libsignal-service-actix/src/push_service.rs

```diff
@@ -170,31 +170,35 @@ impl PushService for AwcPushService {
 
         Self::from_response(&mut response).await?;
 
-        // In order to debug the output, we collect the whole response.
-        // The actix-web api is meant to used as a streaming deserializer,
-        // so we have this little awkward switch.
+        // In order to catch the zero-length output, we have to collect
+        // the whole response. The actix-web api is meant to used as a
+        // streaming deserializer, so we have this little awkward match.
         //
         // This is also the reason we depend directly on serde_json, however
         // actix already imports that anyway.
-        if log::log_enabled!(log::Level::Debug) {
-            let text = response.body().await.map_err(|e| {
-                ServiceError::JsonDecodeError {
-                    reason: e.to_string(),
-                }
-            })?;
-            log::debug!("GET response: {:?}", String::from_utf8_lossy(&text));
-            serde_json::from_slice(&text).map_err(|e| {
-                ServiceError::JsonDecodeError {
-                    reason: e.to_string(),
-                }
-            })
-        } else {
-            response
-                .json()
-                .await
-                .map_err(|e| ServiceError::JsonDecodeError {
-                    reason: e.to_string(),
+        match response.body().await {
+            Ok(text) if text.len() == 0 => {
+                log::debug!("Empty GET response");
+                serde_json::from_slice(b"null").map_err(|e| {
```

Comment:
Maybe we could simplify the `match` by returning the `&str` (so either `text` or `"null"`) and have the `serde_json::from_slice` block a single time? I would also look into refactoring errors that store strings, but that's for another PR I think.

Reply:
It was simpler to return the bytes or error out, and check the length in `from_slice()` instead, but the result is a lot cleaner now! I also opted to return `ServiceError::ResponseError` since it's returned in similar places in hyper.

Reply:
I have refactored this a bit more, let's ship this.